### ISSUE TITLE
Limiting number of allocations when parsing

### ISSUE THREAD

**Issue by vorner** [jbaublitz/neli#52]
Hello

I've noticed the `Vec` type is used quite a lot through the crate, especially when receiving and parsing messages (eg. attributes with payloads that are dependent on the attribute type need to first be parsed as Nlattr<T, Vec<u8>>, which does a heap allocation even if the bit of data is small). Allocating a new vector is not completely free and I'm a bit afraid that in certain workloads (processing large numbers of packets per second on NFLOG or NFQUEUE), this could be a bottleneck.

I have some ideas and I'm willing to put in the effort to implement it, if there's a consensus it is a good idea.

First, the most trivial part ‒ I believe the receive buffer inside NlSocket could be recycled between uses, instead of throwing it away each time. This doesn't change the API and is quite local change.

As for the parsing, I have three variants.

* The simplest, but with the least potential effect. Allowing the usage of SmallVec instead of Vec in places where small buffers are expected. Then, one could parse into a `Vec<Nlattr<T, SmallVec<[u8; 32]>>>` and only the bigger attributes would allocate, + there would be one big overall Vec. Or, alternatively, add support for iterative parsing of `StreamReadBuffer` (this can be done manually now, but is a bit inconvenient to use).
* The `NlAttr<_, Vec<u8>>` has several handy methods. These don't exactly depend on the payload being Vec<u8> specifically, but on `AsRef<[u8]>`. I know `NlAttr<_, &[u8]>` can't be currently parsed by the `Nl` trait, but it could have a dedicated constructor. If one needed it only temporarily (eg. before converting it to the right type), it could help.
* The third would allow full zero-copy parsing support, but is probably the most visible change. It would require changes to the `Nl` trait ‒ similar, how serde's `Deserialize` has a lifetime parameter and can return borrowed values (then it has a `DeserializeOwned` for owned values). But then mostly everything could work with `&[u8]` instead of `Vec<u8>`, or maybe `Cow<[u8]>`. This would allow going as far as the top-level `Nlhdrmsg` borrowing into the buffer living in the `NlSocket` and not allocating or copying anything at all.

What do you think, does anything of it make any sense, or is it just too crazy? I understand all of these add some amount of complexity, so I can accept this being refused ‒ but for me, being able to support higher performance would sound worth it.

**Comment by jbaublitz** [comment]
I'm skeptical of the assertion that this is actually a performance issue. The mistake I often see in development using systems languages that give you the option of allocation to the stack instead of the heap is the idea that heap allocations will always have a noticeable performance impact. Have you done benchmarks or is this simply a concern? I can tell you that memory allocators are often not a bottleneck, particularly when asked to find small contiguous chunks of memory. Reallocations can often be more problematic as they require not only finding a new area of memory but copying all of the existing data over to the new memory block and I was very careful to try to avoid these kinds of reallocations when using the heap. On top of all of this, I've had some of the core developers of Rust mention that using vectors in Rust is heavily optimized to the point where not every creation and drop of a vector directly correspond to an allocation and deallocation. Unless you have specific data backing up the claim that this is indeed a bottleneck, I'm not really willing to implement a drastic change to achieve this.

> First, the most trivial part ‒ I believe the receive buffer inside NlSocket could be recycled between uses, instead of throwing it away each time. This doesn't change the API and is quite local change.

This is not possible. I need the buffer to be the exact size of the number of bytes read from the socket. Because we are making C calls, we cannot rely on the vector to grow and it needs to be big enough to contain the entire message at the time of reading. Once we know how long the message is after reading in the data, we must truncate it to the correct size for parsing. The kinds of size changes make it impossible to simply use the same buffer repeatedly as it will ultimately be a different size for each message. The maximum size required to hold an arbitrary netlink message is large enough that this must be allocated on the heap or it could overflow the stack.

> The simplest, but with the least potential effect. Allowing the usage of SmallVec instead of Vec in places where small buffers are expected. Then, one could parse into a Vec<Nlattr<T, SmallVec<[u8; 32]>>> and only the bigger attributes would allocate, + there would be one big overall Vec. Or, alternatively, add support for iterative parsing of StreamReadBuffer (this can be done manually now, but is a bit inconvenient to use).

I'm mostly interested in your comment about iterative parsing being inconvenient here. While I didn't know about `SmallVec` and am happy to support it somehow in my library if people really want to use that, I'm still not convinced that allocation is the bottleneck. This may or may not happen depending on how easy it is to incorporate, but I'll make a note to investigate that. However, I'm very interested in usability as I think that `neli` definitely has areas that could be improved and I'm open to suggestions!

> The NlAttr<_, Vec<u8>> has several handy methods. These don't exactly depend on the payload being Vec specifically, but on AsRef<[u8]>. I know NlAttr<_, &[u8]> can't be currently parsed by the Nl trait, but it could have a dedicated constructor. If one needed it only temporarily (eg. before converting it to the right type), it could help.

I'm willing to look into this as I've also wondered about how to make this more flexible if desired.

> The third would allow full zero-copy parsing support, but is probably the most visible change. It would require changes to the Nl trait ‒ similar, how serde's Deserialize has a lifetime parameter and can return borrowed values (then it has a DeserializeOwned for owned values). But then mostly everything could work with &[u8] instead of Vec<u8>, or maybe Cow<[u8]>. This would allow going as far as the top-level Nlhdrmsg borrowing into the buffer living in the NlSocket and not allocating or copying anything at all.

I'm not really sure I understand what you mean here. If you read the whole message into a slice buffer, the zero copy aspect is doable in its current form. Take a look at the `buffering` crate for more information on how to deserialize from a slice. If I'm misunderstanding what you mean, could you clarify?

**Comment by vorner** [comment]
> I'm skeptical of the assertion that this is actually a performance issue. The mistake I often see in development using systems languages that give you the option of allocation to the stack instead of the heap is the idea that heap allocations will always have a noticeable performance impact. Have you done benchmarks or is this simply a concern?

For now, this is only a concern, in this specific case. I've seen allocation to be a bottleneck, though and I can try doing some benchmarks later on.

> On top of all of this, I've had some of the core developers of Rust mention that using vectors in Rust is heavily optimized to the point where not every creation and drop of a vector directly correspond to an allocation and deallocation.

Empty ones don't allocate and therefore don't deallocate on drop. But ones that actually hold some data do. This is from the documentation of Vec:

> Vec will allocate if and only if mem::size_of::<T>() * capacity() > 0. 

> This is not possible. I need the buffer to be the exact size of the number of bytes read from the socket.

Let me disagree here a bit. There are two sizes in play ‒ the vector length and vector capacity. Truncate lowers the length, but keeps the extra capacity and therefore allocation. So, it is fine to truncate after reading. As for the exact size needed, it can be changed with `resize_with`. This will use the current allocation as long as the capacity is large enough. So, as long as the requested buffer length doesn't grow since last time (or, more correctly, isn't bigger than maximum of all previous lengths), it doesn't reallocate. Is it OK if I submit a pull request implementing that?

> I'm mostly interested in your comment about iterative parsing being inconvenient here.

I've tried to use something like this:

```rust
while !m.at_end() {
  let attr: NlAttr<u16, Vec<u8>> = Nl::deserialize(m)?;
}
```

This, however, didn't work due to some leftovers at the end of the buffer. The correct way to do that can be seen in the nlattrs.rs:80, which juggles with the size hint. I guess this could be wrapped into some kind of iterator, so one could simply do something like:

```rust
for attr in AttrIter::new(m) {
   // ...
}
```

> I'm not really sure I understand what you mean here.

So, let's say I want to have `NlAttr<u16, &[u8]>` instead of `NlAttr<u16, Vec<u8>>`. I believe I can't implement the `Nl` trait for this type, because now `Self` contains a lifetime needs to be shorter than the lifetime of the `mem` parameter of deserialize. The trait doesn't enforce that.

As a result, I can't really get such type out of `recv_nl`. However, it would be possible to change the trait to `Nl<'b>` and tie the lifetimes together. However, I think this one is too drastic change, because then the bounds will start getting filled with ugly things like `for<'b> Nl<'b>`, etc.

**Comment by jbaublitz** [comment]
I'm still looking this over as I think you make some good points. I want to see if there are areas that could potentially be improved here, but it's taking some time for me to look through your suggestions. I apologize for the delay.

**Comment by jbaublitz** [comment]
I am still thinking through some of this, but I'll respond to the parts I have an answer for.

> For now, this is only a concern, in this specific case. I've seen allocation to be a bottleneck, though and I can try doing some benchmarks later on.

The one area I agree is that libraries that force me into a certain method of allocation are hard to work with. I would prefer to give the user that control if there's a way to do that that makes sense. There may be a few specific cases where this will absolutely not work (I'm thinking of cases where the input size is not known - here there's not a good way to give convenience functions that satisfy all use cases and I plan to ultimately tell people to use the lower level API if they need that much control). Would you object to this from a user perspective - allow that much control from the lower level API and worry less about allocations from the higher level API? I personally think trying to support allocation choices at every level is not really feasible as it is incredibly hard to get any kind of usability at higher levels for more casual users when you have to manage all of the memory yourself at every layer.

> Empty ones don't allocate and therefore don't deallocate on drop. But ones that actually hold some data do. This is from the documentation of Vec:

I may have partially misunderstood here. I'm actually glad I did as I thought the compiler did more magic (I tend to dislike too much of that when code becomes increasingly unrelated to the assembly output). I tested out your claim and it appears the compiler does emit assembly that triggers two allocations when dropping and reallocating a buffer. Good to know!

> Let me disagree here a bit. There are two sizes in play ‒ the vector length and vector capacity. Truncate lowers the length, but keeps the extra capacity and therefore allocation. So, it is fine to truncate after reading. As for the exact size needed, it can be changed with resize_with. This will use the current allocation as long as the capacity is large enough. So, as long as the requested buffer length doesn't grow since last time (or, more correctly, isn't bigger than maximum of all previous lengths), it doesn't reallocate. Is it OK if I submit a pull request implementing that?

You're welcome to submit a PR here but I'm planning on doing some more investigation into how I'm handling allocation at a high level in response to this issue. You make some valid points that I want to seriously consider so it may ultimately not be necessary and might not get merged if I end up going with your `Bytes` suggestion which I will address later.

> I've tried to use something like this:
>
> ```
> while !m.at_end() {
>    let attr: NlAttr<u16, Vec<u8>> = Nl::deserialize(m)?;
> }
> ```
>
> This, however, didn't work due to some leftovers at the end of the buffer.

This is a very good point. Can you open an issue in [buffering](https://github.com/jbaublitz/buffering) so I can make sure I resolve that specific case?

> So, let's say I want to have NlAttr<u16, &[u8]> instead of NlAttr<u16, Vec<u8>>. I believe I can't implement the Nl trait for this type, because now Self contains a lifetime needs to be shorter than the lifetime of the mem parameter of deserialize. The trait doesn't enforce that.
>
> As a result, I can't really get such type out of recv_nl. However, it would be possible to change the trait to Nl<'b> and tie the lifetimes together. However, I think this one is too drastic change, because then the bounds will start getting filled with ugly things like for<'b> Nl<'b>, etc.

Let me just start by saying I'm not opposed to really drastic API changes before v1.0.0. This is the time to get feedback and make sure I'm doing it the best way I possibly can. However, I do think this may not be the ideal suggestion for this use case. This would require keeping around the `StreamReadBuffer`/`StreamWriteBuffer` for the entire lifetime of the parsing. These were intentionally designed to be easily destroyed to extract the internal data and recreated. I don't want them to become long lived. Passing around crucial state for serialization that also contains your data has bitten me before, and I either want to separate out the state and the data into different components or consider an API that uses something more like `Bytes` internally.

What are your thoughts on all of this?

**Comment by vorner** [comment]
> Would you object to this from a user perspective - allow that much control from the lower level API and worry less about allocations from the higher level API?

I think such approach would work for me. I'll just note that it would be great to have an example exercising the low-level API somewhere, both for documentation and to make sure everything works as expected :-).

> You're welcome to submit a PR here but I'm planning on doing some more investigation into how I'm handling allocation at a high level in response to this issue.

In that case I'll probably hold off and see what happens first. I can write it later as well as now.

> This is a very good point. Can you open an issue in buffering so I can make sure I resolve that specific case?

I could if I knew what to ask for there 😇. My original idea was to extract the loop in <Vec<Nlattr<_, _>> as Nl>::deserialize into its own iterator struct and reuse that. That would work without any changes in `buffering`. Basically, something like:

```rust
struct AttrIter<'a, T: Nl, B> {
    mem: &mut StreamReadBuffer<B>,
    size_hint: Option<usize>,
    _t: PhantomData<fn() -> T>,
}

impl<T: Nl, B> Iterator for AttrItem<'_, T, B> {
    type Item = Result<T, DeError>;
    fn next(&mut self) -> Self::Item {
        if self.size_hint > Some(0) || (self.size_hint == None && !self.mem.at_end()) {
            let next = T::deserialize(mem)?;
            if let Some(val) = self.size_hint {
                // TODO: Update size hint
            }
            Some(next)
        } else {
            None
        }
    }
}
```

(This probably won't compile for some little annoying details…)

Then the above deserialize could be rewritten as `AttrIter::new(mem).collect()`, but it would still be usable as a separate iterator.

**Comment by jbaublitz** [comment]
@vorner I've finally gotten around to looking into `Bytes`. Here are my thoughts.

Overall, I think it could be very helpful and I think the main thing we could do here is provide a single `Bytes` allocation of size `MAX_NL_LENGTH`. `Bytes` is backed by a vector so this would be very little overhead to do a single allocation and then reuse that buffer for operations later. This also allows a pretty nice interface for subslices.

I don't think this will be entirely possible, however, without some changes to `buffering`. I still want to address #50 around skipping messages or rewinding if something fails to deserialize. If we store the entire buffer as a `Bytes` object, this still means we need to know the beginning and end indexes if we intend to provide a `Bytes` object as a subslice. This requires some bookkeeping by `StreamReadBuffer`. We can easily expose a nicer interface that returns either the same subslice `Bytes` object or the next one, but that does still require some internal tracking of position. Does that make sense?

**Comment by vorner** [comment]
> I don't think this will be entirely possible, however, without some changes to buffering. 

Alternative is using combination of Cursor<Bytes>, which implements the [`Buf`](https://docs.rs/bytes/0.4.12/bytes/buf/trait.Buf.html) trait, which is very similar to what `StreamReadBuffer` does (there's similar BufMut for writing). With clever splitting and adjusting the lenghts one could implement the size hints too.

**Comment by jbaublitz** [comment]
So here are my concerns with that approach:
1. The `Buf` trait does not appear to support native endianness as byteorder does. This is a significant problem as I want neli to be supported for both big and little endian which would get incredibly messy without some kind of option to use the default endianness of the system.
2. Can you propose an interface at the `Nl` level for how this would look? I'm not really on board with the idea of making `Nl` more complicated and pushing the simplicity up a level to the level of iterators. If there are going to be messy implementation details around slicing, I want those abstracted away from `Nl` and simply make changes to buffering. I want `Nl` implementations to look as simple as possible.
3. If you're convinced you can keep `Nl` simple with your outlined approach, please provide a small code snippet to demonstrate how this would work as, from my perspective, this sounds like the user would have to worry about slicing the `Bytes` struct into smaller slices and passing them along to serialize methods called within the current serialize method and I do not want them to have to worry about that. I'd rather rework bufffering to allow calling a method that simply takes a length and returns a new `StreamReadBuffer` with an underlying `Bytes` subslice pointing to the original.

**Comment by vorner** [comment]
Hello

I'm not sure my ideas satisfy the simplicity criterion, but I'd have a sketch of some ideas.

I think the endians is not a *big* problem. It does export LittleEndiand and BigEndian and we could create a type alias for either of them with `#[cfg]` attribute and re-export it. And we could add an extension trait with `get_u16` to it, or maybe ask upstream to add them.

Second, the basic method for deserialization would be something like:

```rust
trait Nl {
    fd deserialize(buf: Bytes) -> Result<Self, Error>;
    ...
}
```

This one would necessarily take care of all the complexities of handling slices and whatever. However, most types around netlink seem to be some kind of fixed header + bunch of attributes. So we could have another trait named `NlStruct`, which could look something like this:

```rust
trait NlStruct {
    const HDR_SIZE: usize;
    fn deserialize_struct(header: Cursor<Bytes>, attributes: AttrIterator) -> Result<Self, Error>;
    fn serialize_struct(&self, output: &mut Cursor<Vec<u8>>) -> Result<(), Error>;
```

* The `Nl` trait would then be implemented by means of this trait automatically, eg. `for<S: NlStrung> impl Nl for S { ... }`, which would do all the slicing and such.
* The header would be guaranteed to be exactly `HDR_SIZE` bytes long, so the implementer of `NlStruct` would not have to take care of range checking and would just deserialize all the fields, then go through the attributes and handle them.
* The iterator would produce tuples `(u16, Bytes)`, the left the indentifier, the right the slice of data of the attribute. Alternatively, the NlStruct trait could have a type parameter and specify what type the identifier is instead of u16.

So, the implementation of the `LogPacket` I've added recently would look something like this (I'm omitting some fields for brevity):

```rust
fn deserialize(mut header: Cursor<Bytes>, attrs: AttrIter) -> Result<Self, Error> {
    let result = Self::dummy_instance();
    result.hw_protocol = header.get_u16_be();
    result.hook = header.get_u16(); // Done with the extension trait
    header.get_u8(); // skip pad
    ...
    for (id, attr) in attrs {
         match NlLogAttr::from(id) {
              NlLogAttr::Mark => result.mark = u16::decode(attr)?,
              NlLogAttr::Timestamp => result.timestamp = Timestamp::decode(attr)?,
              ...
         }
    }
    Ok(result)
}
```

I believe such code would be reasonably readable and intuitive and I think the (more complex) low level deserialize would need to be implemented only in rare circumstances (and in the in-crate case for the NlStruct case).

How does that look to you?

**Comment by jbaublitz** [comment]
> I think the endians is not a big problem. It does export LittleEndiand and BigEndian and we could create a type alias for either of them with #[cfg] attribute and re-export it. And we could add an extension trait with get_u16 to it, or maybe ask upstream to add them.

I'm going to have to say no here. This provides no obvious benefit that I can see over the use of byteorder, and would even introduce more complexity into my code. I can even see using `Bytes` and `Cursor` together, which was what I was considering adding into buffering based on your suggestion, and still using one of byteorder's provided traits to read out the integers. This is not an allocation issue when dealing with integer byte order and the byteorder crate already supports this nicely. No need to reinvent the wheel.

> I believe such code would be reasonably readable and intuitive and I think the (more complex) low level deserialize would need to be implemented only in rare circumstances (and in the in-crate case for the NlStruct case).

I've taken a look at the code samples and I have to say that while the `Nl` trait definitely does need some adjustment, I'd hesitate to go with your solution as I'd rather make `Nl` better an have the buffering crate become more helpful instead of creating more traits that would simply make this more confusing. If we specialize a trait for each use case, netlink is so hard to predict and contains so many different formats that can get deserialized that I have no doubt I will be left maintaining a large number of traits that I'm not sure will actually make this library more intuitive to use.

What I do like about your suggestion is providing some kind way to rewind and retry, and to fast-forward to the next message. I also am pretty happy with what I've seen of the `Bytes` struct and the suggestion that we don't reallocate the buffer in the socket on every read. All of these are helpful and I think I'm going to put together a PR targeting `v0.5.0` for you to look over and give last comments on before I merge it into the `v0.5.0` branch.

**Comment by vorner** [comment]
OK, I was just providing possible alternatives :-).

**Comment by jbaublitz** [comment]
@vorner Hello! I haven't forgotten about this. I've been trying to prioritize this for new development as I think it's going to be the most complicated change from an API perspective. What I've been doing is working on the buffering crate, and on the `v0.5.0-dev` branch, I have the beginning of a stateless read method which will actually take over most of the work of deserialization from neli. It's pushed to that branch if you're interested in taking a look at the preliminary work (granted it's still subject to change and has not yet been tested).

The benefits here should be as follows:
* This may eventually remove mutability requirements by passing owned state and threading it through the read methods.
* This will provide much more granular control over determining whether a type has been properly serialized based on the total length of the packet and each sub-structure as you suggested.
* It will use `Bytes` and `BytesMut` containers decrease allocations.

**Comment by jbaublitz** [comment]
Hi @vorner. You were absolutely right here! Upon first glance, it makes a bit more sense to place the parsing logic in neli as opposed to buffering. I was doing some exploration in buffering and I came up with a few ideas for how to make parsing easier and I'm going to use that moving forward. I wasn't thinking it through, however, and bumped into the problem with `impl Trait` on arbitrary types. There's not a trait-friendly way to implement this in a separate crate so it seems to be a bit challenging to make a flexible way to serialize/deserialize primitives in the standard library. The only way I can think of doing that in a more generic way is to essentially reimplement serde which doesn't make sense. As a result, I think it makes less sense to try to move the logic into buffering. My current thought is to provide macros to deal with driving the state as I tried to do in buffering and change the `Nl` trait slightly to support that. Ultimately it'll boil down to your original suggestion of clever slicing but with some macros to make it more user friendly so it both implements your suggestion and satisfies my requirements. Thank you for your input!

**Comment by vorner** [comment]
Hello

Sorry for taking that long to have a look and read your answer. I guess macros are fine, but it would be great if even the manual way without macros would be somewhat sane, even if eg. more verbose.

**Comment by jbaublitz** [comment]
I don't think this will be a concern in how I'm implementing the neli implementation. You get a raw `Bytes` or `BytesMut` object in the `Nl` traits moving forward as of v0.5.0 and you can choose to parse that any way you like (including using a `Cursor` in only one `serialize` implementation for example) as long as the contract of passing in a `Bytes`/`BytesMut` object of the exact length of the binary representation is upheld. This gives maximum flexibility for those who want to do more complicated parsing while the default is supported using relatively simple macro syntax.

**Comment by vorner** [comment]
When playing with the library, I've discovered few more places that would benefit from lesser amount of Vec's around. I believe these would be more easy / local fixes than the above and I'm willing to send pull requests if you agree with them.

* Flags usage. Eg the [`Nlmsghdr`](https://docs.rs/neli/0.4.3/neli/nl/struct.Nlmsghdr.html) needs a flags field, specified as a vec. One can reuse the Nlmsghdr when multiple messages are to be sent to the kernel, but it feels a bit unnatural to modify it, and creating a new one every time „eats“ the vec. Eventually, this'll go over the fields and OR them together to form one `c_int`, which at least *feels* wasteful. Would using bitflags make more sense? Eg. one would call `Nlmsghdr::new(None, SomeType, NlmF::Request | NlmF::Ack], ...)` instead the same with `vec![NlmF::Request, NlmF::Ack]`.
* Changing the `impl<'a, T> Nl for &'a [Nlattr<T, Vec<u8>>]` to `impl<'a, T, P> Nl for &'a [Nlattr<T, P>]` which would allow better flexibility overall, but also make it possible for have borrowed P. I think this is even API compatible change, because it just adds more available impls and doesn't modify existing ones.

**Comment by jbaublitz** [comment]
This should be resolved in `v0.5.0-dev`. I'm going to do additional work on this at the feature level so I'll leave this open. Hopefully the result will be three levels of heap allocations:
* `prefer-heap`
* `prefer-stack`
* `disable-heap`

**Comment by jbaublitz** [comment]
I've removed all of the allocations that I can at the moment. Closing.
